tofu/tests/tests01_geom/test_04_sampling.py

```diff
@@ -0,0 +1,271 @@
+def test25_sa_integ_map(ves_poly=VPoly, debug=1):
+    import tofu.geom as tfg
+    import matplotlib.pyplot as plt
+
+    ves_norm = compute_ves_norm(ves_poly)
+
+    ves = tfg.Ves(
+        Name="DebugVessel",
+        Poly=ves_poly,
+        Type="Tor",
+        Exp="Misc",
+        shot=0
+    )
+
+    config = tfg.Config(Name="test",
+                        Exp="Misc",
+                        lStruct=[ves])
+
+    part = np.array([[3.5, 0, 0], [10, 10, 10]], order='F').T
+    part_rad = np.r_[0.1, 0.1]
+    rstep = zstep = 0.01
+    phistep = 0.5
+    limits_r, limits_z = compute_min_max_r_and_z(ves_poly)
+
+    res = GG.compute_solid_angle_map(part, part_rad,
+                                     rstep, zstep, phistep,
+                                     limits_r, limits_z,
+                                     ves_poly=ves_poly,
+                                     ves_norm=ves_norm,
+                                     )
+    pts, sa_map, ind, rdrdz = res
+
+    # check sizes
+    npts_ind = np.size(ind)
+    npts_vol = np.size(rdrdz)
+    dim, npts = np.shape(pts)
+    npts_sa, sz_p = np.shape(sa_map)
+
+    if debug > 0:
+        print(f"sa_map is of size {npts_sa},{sz_p}")
+    assert dim == 2
+    assert sz_p == np.shape(part)[1]
+    assert npts == npts_sa == npts_vol
+
+    # ... Testing with exact function .........................................
+    res = GG._Ves_Vmesh_Tor_SubFromD_cython(rstep, zstep, phistep,
+                                            limits_r, limits_z,
+                                            out_format='(R,Z,Phi)',
+                                            margin=1.e-9)
+
+    pts_disc, dvol, ind, reso_r, reso_z, reso_phi = res
+    npts_disc = np.shape(pts_disc)[1]
+    sang = config.calc_solidangle_particle(pts_disc,
+                                           part,
+                                           part_rad,
+                                           approx=False)
+
+    if debug > 0:
+        fig = plt.figure(figsize=(14, 8))
+        ax = plt.subplot(121)
+        ax.plot(pts_disc[0, :], pts_disc[1, :], '.b')
+        fig.suptitle("testing still")
+        plt.savefig("mocomoco")
+
+    assert (sz_p, npts_disc) == np.shape(sang)
+    assert npts_disc > npts
+
+    sz_z = np.abs(limits_z[1] - limits_z[0]) // reso_z
+    sz_r = np.abs(limits_r[1] - limits_r[0]) // reso_r
+    assert npts_sa == sz_z * sz_r
+
+    sa_map_py = np.zeros((npts_sa, sz_p))
+
+    for ii in range(npts_disc):
+        i_r = int((pts_disc[0, ii] - limits_r[0]) // reso_r)
+        i_z = int((pts_disc[1, ii] - limits_z[0]) // reso_z)
+        ind_pol = int(i_r * sz_z + i_z)
+        if ind_pol == 39999:
+            print("solid angle = ", sang[0, ii], reso_phi[i_r])
+        for pp in range(sz_p):
+            sa_map_py[ind_pol, pp] += sang[pp, ii] * np.pi * reso_phi[i_r]
+
+    print("reconstructed sa map =", sa_map_py)
+    print("cython computed = ", sa_map)
+    assert np.allclose(sa_map_py, sa_map)
```

Comment:
This is where unit tests fail right ?
Maybe try on a smaller set of point for easier debugging ?